presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcFileWriter.java

```diff
@@ -191,11 +223,60 @@ public long getValidationCpuNanos()
         return validationCpuNanos;
     }
 
+    public void setMaxWriteId(long maxWriteId)
+    {
+        this.maxWriteId = OptionalLong.of(maxWriteId);
+    }
+
     @Override
     public String toString()
     {
         return toStringHelper(this)
                 .add("writer", orcWriter)
                 .toString();
     }
+
+    private Block[] buildAcidColumns(Block rowBlock, AcidTransaction transaction)
+    {
+        int positionCount = rowBlock.getPositionCount();
+        int bucketValue = computeBucketValue(bucketNumber.isPresent() ? bucketNumber.getAsInt() : 0, 0);
+        // operation, originalWriteId, bucket, rowId, currentWriteId, row<>
+        return new Block[] {
+                RunLengthEncodedBlock.create(INTEGER, (long) getOrcOperation(transaction), positionCount),
+                RunLengthEncodedBlock.create(BIGINT, transaction.getWriteId(), positionCount),
+                RunLengthEncodedBlock.create(INTEGER, (long) bucketValue, positionCount),
+                buildAcidRowIdsColumn(positionCount),
+                RunLengthEncodedBlock.create(BIGINT, transaction.getWriteId(), positionCount),
+                rowBlock
+        };
+    }
+
+    private int getOrcOperation(AcidTransaction transaction)
+    {
+        switch (transaction.getOperation()) {
+            case INSERT:
+                return 0;
+            default:
+                throw new IllegalStateException("In getOrcOperation, transaction isn't INSERT, transaction " + transaction);
+        }
+    }
+
+    private Block buildAcidRowIdsColumn(int positionCount)
+    {
+        long[] rowIds = new long[positionCount];
+        for (int i = 0; i < positionCount; i++) {
+            rowIds[i] = i;
+        }
+        return new LongArrayBlock(positionCount, Optional.empty(), rowIds);
+    }
+
+    public static int extractBucketNumber(int bucketValue)
+    {
+        return (bucketValue >> 16) & 0xFFF;
+    }
+
+    public static int computeBucketValue(int bucketId, int statementId)
+    {
+        return 1 << 29 | bucketId << 16 | statementId;
```

Comment:
Should we add checks that statmentId is at most 2^16 and bucketId is at most 2^12 (I guess that is the limit based on 0xfff above)

Reply:
Good idea; added.